src/VisualStudio/Core/Def/TaskList/ExternalErrorDiagnosticUpdateSource.cs

```diff
@@ -139,385 +111,170 @@ public ImmutableArray<DiagnosticData> GetBuildErrors()
         {
-            // nothing to do
-            if (inProgressState == null)
-            {
-                return;
-            }
-
-            // Mark the status as updated to refresh error list before we invoke 'SyncBuildErrorsAndReportAsync', which can take some time to complete.
-            OnBuildProgressChanged(inProgressState, BuildProgress.Updated);
-
-            // We are about to update live analyzer data using one from build.
-            // pause live analyzer
-            using var operation = _notificationService.Start("BuildDone");
-            if (_diagnosticService is DiagnosticAnalyzerService diagnosticService)
-                await SyncBuildErrorsAndReportOnBuildCompletedAsync(diagnosticService, inProgressState).ConfigureAwait(false);
-
-            // Mark build as complete.
-            OnBuildProgressChanged(inProgressState, BuildProgress.Done);
-        }, GetApplicableCancellationToken(inProgressState));
+            await ClearPreviousAsync(projectId: null, cancellationToken).ConfigureAwait(false);
+        });
     }
 
     /// <summary>
-    /// Core method that de-dupes live and build diagnostics at the completion of build.
-    /// It raises diagnostic update events for both the Build-only diagnostics and Build + Intellisense diagnostics
-    /// in the error list.
+    /// Called serially in response to the sln build UI context completing.
     /// </summary>
-    private ValueTask SyncBuildErrorsAndReportOnBuildCompletedAsync(DiagnosticAnalyzerService diagnosticService, InProgressState inProgressState)
+    internal void OnSolutionBuildCompleted()
     {
-        var solution = inProgressState.Solution;
-        var cancellationToken = inProgressState.CancellationToken;
-        var (allLiveErrors, pendingLiveErrorsToSync) = inProgressState.GetLiveErrors();
-
-        // Raise events for build only errors
-        using var argsBuilder = TemporaryArray<DiagnosticsUpdatedArgs>.Empty;
-        var buildErrors = GetBuildErrors().Except(allLiveErrors).GroupBy(k => k.DocumentId);
-        foreach (var group in buildErrors)
-        {
-            cancellationToken.ThrowIfCancellationRequested();
-
-            if (group.Key == null)
-            {
-                foreach (var projectGroup in group.GroupBy(g => g.ProjectId))
-                {
-                    Contract.ThrowIfNull(projectGroup.Key);
-                    argsBuilder.Add(CreateArgsToReportBuildErrors(projectGroup.Key, solution, projectGroup.ToImmutableArray()));
-                }
-
-                continue;
-            }
-
-            argsBuilder.Add(CreateArgsToReportBuildErrors(group.Key, solution, group.ToImmutableArray()));
-        }
-
-        ProcessAndRaiseDiagnosticsUpdated(argsBuilder.ToImmutableAndClear());
-
-        // Report pending live errors
-        return diagnosticService.SynchronizeWithBuildAsync(_workspace, pendingLiveErrorsToSync, onBuildCompleted: true, cancellationToken);
+        _ = ClearInProgressState();
```

Comment:
yes.  this now seems entirely racey *if* the ABWQ also accesses whatever fields this modifies.  Would it not be better to put this state processing intot eh queue, or at least capture what we so that the queue operation operates on the right state of this type?

Reply:
resolved offline (state not used)